static/js/modal.js

```diff
@@ -0,0 +1,86 @@
+var Modal = (function () {
+    var funcs = {
+        createModal: function () {
+            var overlay = document.createElement("overlay");
+            overlay.className = "overlay new";
+            overlay.setAttribute("data-overlay", "custom-modal-" + Math.random().toString(16));
+
+            var modal = document.createElement("div");
+            modal.className = "zulip-modal";
+
+            overlay.appendChild(modal);
+
+            return {
+                overlay: overlay,
+                modal: modal,
+            };
+        },
+    };
+
+    var __Modal = function (html) {
+        this.meta = funcs.createModal();
+        this.meta.modal.innerHTML = html;
+        this.meta.hasMounted = false;
+    };
+
+    __Modal.prototype = {
+        html: function (html) {
+            this.meta.modal.innerHTML = html;
+
+            return this;
+        },
+
+        get: function () {
+            return this.meta.overlay;
+        },
+
+        addClass: function (className) {
+            this.get().classList.add(className);
+        },
+
+        show: function () {
+            var overlay = this.get();
+            document.body.appendChild(overlay);
+
+            setTimeout(function () {
+                overlays.open_overlay({
+                    name: overlay.getAttribute("data-overlay"),
+                    overlay: $(overlay),
+                    on_close: function () {
+                        hashchange.exit_overlay();
+                    },
+                    allowOverlayOverlap: true,
+                });
```

Comment:
Why do we want to create another mechanism for overlays? IMO it's better to inherit the base properties from the overlay component than to reinvent the modal again.

The reason it has to integrate with the modals is because we use the same HTML shell as the other modals in order to correctly use the JS events associated with the modals that already exist (e.g. click on a modal overlay directly to close it).